Feature/8498 sc main vm wizard

malinajirka opened the pull request:
Fixes #8498 

Note: Change the target branch to `feature/master-site-creation` as soon as #8619 is merged.

Introduces `SiteCreationMainViewM` which is responsible for storing state and transition between screens in the Site Creation flow.

Main changes
1. NewSiteCreationActivity converted to kotlin + cleaned up.
- I didn't want to do it in this PR, but I had some troubles with lint when I used a Kotlin enum in a Java class + I needed to use an extension function.

2. LiveDataUtils - observeEvent
- We use `SingleLiveEvent` for cases in which we want to notify an observer just once - for example when we want to show a snackbar message. However, `SingleLiveEvent` extends `MutableLiveData` and I needed to use `MediatorLiveData`. 
-> Vojta has introduced this approach in #8566. I'm still not sure this is the best approach, but I see it's advantages. I decided to copy-paste it until we make a final decision whether we'll use `observeEvent` approach everywhere or we'll try to create `SingleLiveEventMediator` (which won't be that simple as LiveDataMediator uses internal api).

3. `NewSiteCreationVerticalsResultObservable` and `NewSiteCreationSegmentsResultObservable`
- This is something I'm still not sure about. I've introduced two ResultObservables which I use to propagate results from a Verticals/Segments fragments to the `NewSiteCreationActivity`. I'd like the approach if we had `ViewModelScope` in Dagger, but using singletons feels a bit hacky. On the other hand I think it's still better than using regular `(XyZListener) getActivity()` approach.  Wdyt?

4. Fixed few issues
 - Segments/Verticals Fragments were passing activity to the ViewModel provider
- fixes skip button and verticals item click listeners

5. WizardManager + WizardStep + WizardState
- I've created a robust/over-enginnered solution (with WizardStepFactories etc), but I decided to remove it and use this simple approach instead as it's way easier to use and understand + we don't need it at the moment.

Note:
We don't save the state into SavedState at the moment. This is something I want to introduce in one of the following PRs as it handles just an edge case scenario in which the app is killed while in background. #8683


To test:
1. My Site
2. Click on the Switch Site button
3. Click on the plus sign icon 
4. Select Create WordPress.com site
5. Test the flow - go forth and back  and rotate your device (The domains screen is not ready yet, I've added it just so we have a third step)

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/util/wizard/WizardManager.kt:
```diff
@@ -0,0 +1,52 @@
+/**
+ * Marker interface representing a single step/screen in a wizard
+ */
+interface WizardStep
```
I am only mentioning this because we talked about this on the call. An alternative approach to keeping an index might be to use an approach similar to a LinkedList or a tree. Linked list approach might look something like this:

```
interface WizardStep {
    val previous: WizardStep?
    val next: WizardStep?
}
```

In this ^ case, `WizardManager` would take the initial step instead of the list. These changes would allow us to get rid of index tracking and although unlikely, dealing with out of bounds indexes.

_I have a sneaking suspicion that there might be a nice way to tie this to the communication between view models such as `onVerticalSelected` without actually exposing that it's a step to them._

malinajirka replied:
👍 Thanks for writing this down. I've decided to think about it for a few more days as it's not urgent and the current solution works.

oguzkocer replied:
Sounds good!

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/util/LiveDataUtil.kt:
```diff
@@ -0,0 +1,20 @@
+    })
+}
+
+open class Event(var hasBeenHandled: Boolean = false)
```
We already talked about this on the call, but just adding a note that if we end up keeping the code, it'd be good to improve this. I think we could make this an interface, but if it has to be a class, let's at least make it `abstract`. Also, let's not expose the `hasBeenHandled` and instead just have a method such as `handle` and keep the state internal.

malinajirka replied:
I've removed it. However, even if we decide to keep it after all, I'd suggest improving it in another PR as it would create merge conflicts with Vojta's branch.

oguzkocer replied:
Yeap, that makes sense. If that turns out to be the case, do you mind opening an issue for that so we don't forget about it?

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/verticals/NewSiteCreationVerticalsViewModel.kt:
```diff
@@ -269,14 +280,14 @@ class NewSiteCreationVerticalsViewModel @Inject constructor(
     }
 
     sealed class VerticalsListItemUiState {
+        lateinit var onItemTapped: () -> Unit
```
I feel like using `lateinit` for this is kind of error prone. I'd personally prefer either `nullable` or a different approach.

malinajirka replied:
Good point! Using a nullable type with a unittest which checks it's not null is safer.

malinajirka commented:
Thanks @oguzkocer. I've replaced the NewSiteCreationVerticalsResultObservable and NewSiteCreationSegmentsResultObservable with regular approach with listeners as we discussed. (The NewSiteCreationActivity implements interfaces and the Segments/Verticals fragments cast the activity to that interface and when a segment/vertical is selected they pass the data to the listener.)

I've also fixed all the issues you've pointed out. It's ready for another pass. Thanks!

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/viewmodel/SingleEventObservable.kt:
```diff
@@ -0,0 +1,32 @@
+            throw IllegalStateException("SingleEventObservable can be observed only by a single observer.")
+        }
+        sourceLiveData.observe(owner, Observer {
+            if (it !== lastEvent) {
```
I am not sure if instance comparison is exactly what we want here. I know we don't want a value comparison, because it's totally valid to, for example, show the same toast message multiple times such as after a pull to refresh. I am skeptical about the instance creation as well, because it looks like same event can be triggered multiple times as different instances and we end up having to handle it when we don't want to. Please check out main review comment where I'll share more details about my testing experience which is related to this.

One ugly alternative is to take an equals function in the constructor so we can let the view models decide how each event should be compared. For this PR, it'll probably mean that we'll compare the step or maybe step and state, but for a toast it might be an instance comparison.

Not being able to follow up the pending approach feels really bad, because ideally, it should work just like `SingleLiveEvent` without view model having to provide more logic. Let's maybe discuss this further on Monday after we can think about it a bit further.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/verticals/NewSiteCreationVerticalsViewHolder.kt:
```diff
@@ -21,14 +22,17 @@ sealed class NewSiteCreationVerticalsViewHolder(internal val parent: ViewGroup,
             divider.visibility = if (uiState.showDivider) View.VISIBLE else View.GONE
-            // TODO add onClick listener
+            container.setOnClickListener {
+                uiState.onItemTapped?.invoke() ?: AppLog.e(AppLog.T.MAIN, "onClickListener method reference is null.")
```
Should we require `onItemTapped` to be not null at this point to follow "your" 😁 fail fast principle? My suggestion for using a `nullable` wasn't to ignore the error, but make it more obvious because I feel like `lateinit`s error is not as helpful as having a `nullable` and expecting it to be set at a certain point. Having said that I am totally fine with this approach as well, I think both approaches are better than `lateinit` for a callback, so it's your call!

malinajirka replied:
:D yep, you are probably right.... I felt like in this case it's not so crucial as it's pretty obvious it's a bug when nothing happens after you click on a list item... but we might not know about it until someone contacts support, so it's better to just crash.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/verticals/NewSiteCreationVerticalsViewHolder.kt:
```diff
@@ -53,7 +57,7 @@ sealed class NewSiteCreationVerticalsViewHolder(internal val parent: ViewGroup,
             retry.text = itemView.context.getText(uiState.retryButtonResId)
             itemView.setOnClickListener {
-                uiState.onItemTapped.invoke()
+                uiState.onItemTapped?.invoke() ?: AppLog.e(AppLog.T.MAIN, "onClickListener method reference is null.")
```
Same comment about `onItemTapped`, just putting this here as a reminder.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/verticals/VerticalsScreenListener.kt:
```diff
@@ -0,0 +1,5 @@
+package org.wordpress.android.ui.sitecreation.verticals
+
+interface VerticalsScreenListener {
```
I think it'd be worth having a separate `onSkip` method here and making the `verticalId` one not nullable. Ideally these methods should do only one thing.

malinajirka replied:
Yes, I had this feeling too. I created a new interface so we don't have bunch of `onSkip()` methods.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/NewSiteCreationActivity.kt:
```diff
@@ -0,0 +1,90 @@
+
+    private fun observeVMState() {
+        mainViewModel.navigationTargetObservable
+                .observe(this, Observer { target -> target?.let { showStep(target) } })
```
Just sharing my preference: I like to start with `it` in the observer and only assign the name after the null check. It avoids having multiple `target`s or using the `it` at the final stage where it's less meaningful. No change necessary, just sharing :)

```suggestion
                .observe(this, Observer { it?.let { target -> showStep(target) } })
```

malinajirka replied:
>It avoids having multiple targets or using the it at the final stage where it's less meaningful.

Hmm, but I have just a single target and I don't use `it` at all, right? Smartcast takes care of casting T? -> T. I like this approach a bit more... but I don't feel strong about it.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/NewSiteCreationActivity.kt:
```diff
@@ -0,0 +1,90 @@
+            onBackPressed()
+            return true
+        }
+
```
👀 🙈

malinajirka replied:
:D This is not a stray empty line -> it separates the end of the if. But I agree it doesn't add much value:P.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/NewSiteCreationMainVM.kt:
```diff
@@ -0,0 +1,61 @@
+typealias NavigationTarget = WizardNavigationTarget<SiteCreationStep, SiteCreationState>
+
+class NewSiteCreationMainVM @Inject constructor() :
+        ViewModel() {
```
Nitpick: Can we ⬆️ this line?

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/NewSiteCreationMainVM.kt:
```diff
@@ -0,0 +1,61 @@
+        ViewModel() {
+    private val wizardManager: WizardManager<SiteCreationStep> = WizardManager(
+            // TODO we'll receive this from a server/Firebase config
+            Arrays.asList(
```
Nitpick: We can use `listOf` instead.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/NewSiteCreationMainVM.kt:
```diff
@@ -0,0 +1,61 @@
+    private var isStarted = false
+    private var siteCreationState = SiteCreationState()
+
+    val navigationTargetObservable: SingleEventObservable<NavigationTarget> = SingleEventObservable(
```
I have played with this once again with a different approach and the more I think about it and play with it, the more I feel like that the `LiveData` is really not right tool to handle single events like navigation, toast etc. We should look into alternatives for it as a hack week project or something.

malinajirka replied:
Yes, I'm starting to have the same feeling. We've tried several different approaches and all of them have some down-sides. Unless we come up with an ultimate solution, I think we should stop using LiveData for such cases.

malinajirka replied:
I've discussed it with Vojta and we all have the same feeling. We are thinking about creating a custom LifeCycleAware Listener/Observable which would solve this issue once and for all.

malinajirka commented:
Thanks @oguzkocer;). I've fixed all the issues.

>In the last step, repeatedly tapping on the back button will also result with the crash below. This was one of my concerns with the check that results in crash, because with events, it is always possible to get a callback when we don't expect to. I really think we should ignore them or deal with them somehow instead of crashing it. I know this one is an edge case and we'll fix it, but I am not sure if there won't be an edge case that's totally valid and end up resulting in a crash.

I agree that this exception doesn't make any sense and is actually incorrect -> as you never know whether and when an event from a canceled coroutine arrives.
However, I'd like to keep the exception in the `FetchSegmentsPrompt.fetchSegmentsPrompt` and `FetchSegmentsUseCase.fetchCategories` methods as it's clients(VMs) responsibility to call the method only when there is no request in progress.

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/NewSiteCreationMainVM.kt:
```diff
@@ -0,0 +1,61 @@
+        wizardManager.showNextStep()
+    }
+
+    fun onSkipClicked() {
```
I feel like we use `clicked` everywhere but `onBackPressed`. I think the OS uses the same terminology, so I'd rather stick to it, if you don't mind.

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/segments/NewSiteCreationSegmentsFragment.kt:
```diff
@@ -69,7 +80,7 @@ class NewSiteCreationSegmentsFragment : NewSiteCreationBaseFormFragment<NewSiteC
             }
         })
-
+        viewModel.segmentSelected.observe(this, Observer { it?.let { segmentsScreenListener.onSegmentSelected(it) } })
```
>Nitpick: There is an AS warning about the last it being shadowed.

This is already fixed in one of the working branches.

>Also completely unrelated to this PR, in initRetryButton, view is not used in setOnClickListener, we could get rid of it.

I'll fix this one asap :)
